### PR TITLE
Complete implementation of set wifi if not exist

### PR DESCRIPTION
**Pull request by 0xdev22**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,6 +25,8 @@ For the management calls, the process needs to have the `CAP_NET_BIND_SERVICE` c
     * _Read._ Return the ASCII-encoded Signal Strength of the Unit in dBm
   * Get Wifi connection status characteristic `5c305a15-6859-4d6c-a87b-8d2c98c9f6f0`
     * _Read._ Return the ASCII-encoded SSID of current wifi connected to the unit
+  * Set Wifi Connection characteristic `5c305a16-6859-4d6c-a87b-8d2c98c9f6f0`
+    * _Write._ Send in wifi SSID and Password of wifi to connect to
```
Comment (elffjs) [inline README.md:29]: Once the API settles, might we describe what exactly one submits and in what order here?

File: main.go [hunk 2]
```diff
@@ -533,6 +594,101 @@ func main() {
...
+		}
+
+		if SSID == "" {
+			SSID = string(value)
+		} else {
+			WifiPasswordKey = string(value)
+		}
+
+		log.Println(SSID, "SSID", WifiPasswordKey, "key")
+		if SSID != "" && WifiPasswordKey != "" {
```
Comment (elffjs) [inline main.go:628]: How about we go with Yev's simpler option: just replace the entire list with our new network? No fiddling with priority and so on.

If you're okay with this I'll modify the ticket.

File: main.go [hunk 3]
```diff
@@ -533,6 +594,101 @@ func main() {
...
+			var er error
+			wifiCnxnLst, er := getAvailableWifiConnections(unitID, SSID)
+			if er != nil {
+				log.Fatalf("Failed to set wifi connection: %s", er)
```
Comment (elffjs) [inline main.go:632]: I would avoid calling `Fatal` in any of these handlers.

File: main.go (hunk 2 above)
Comment (0xdev22) [reply main.go:628]: Yeah that works

File: main.go [hunk 4]
```diff
@@ -566,6 +632,72 @@ func main() {
...
+				log.Printf("Error setting wifi connection: %s.", err)
+			}
+		}()
+
+		var s string
+		err = json.Unmarshal(value, &s)
+		if err != nil {
+			log.Printf("Error setting wifi: %s", err)
+			return
+		}
```
Comment (elffjs) [inline main.go:657]:
```suggestion
```
I think you can lose this.

File: main.go [hunk 5]
```diff
@@ -566,6 +632,72 @@ func main() {
...
+
+		var req setWifiRequest
+		err = json.Unmarshal([]byte(s), &req)
+		if err != nil {
+			log.Printf("Error setting wifi: %s", err)
```
Comment (elffjs) [inline main.go:662]:
```suggestion
			log.Printf("Error unmarshaling wi-fi payload: %s", err)
```
Just to be more specific.

File: main.go [hunk 6]
```diff
@@ -242,6 +274,40 @@ func getWifiStatus(unitID uuid.UUID) (connectionObject wifiConnectionsResponse,
 	return
 }
 
+func getAvailableWifiConnections(unitID uuid.UUID, ssID string) (connections []wifiEntity, err error) {
```
Comment (elffjs) [inline main.go:277]: I think we can delete this function.

File: main.go [hunk 7]
```diff
@@ -59,12 +63,19 @@ const (
 var lastSignature []byte
 var lastVIN string
 
+var SSID string
+var WifiPasswordKey string
```
Comment (elffjs) [inline main.go:67]:
```suggestion
```

**Review by elffjs — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -25,6 +25,8 @@ For the management calls, the process needs to have the `CAP_NET_BIND_SERVICE` c
     * _Read._ Return the ASCII-encoded Signal Strength of the Unit in dBm
   * Get Wifi connection status characteristic `5c305a15-6859-4d6c-a87b-8d2c98c9f6f0`
     * _Read._ Return the ASCII-encoded SSID of current wifi connected to the unit
+  * Set Wifi Connection characteristic `5c305a16-6859-4d6c-a87b-8d2c98c9f6f0`
+    * _Write._ Send in wifi SSID and Password in the json string form e.g. "{\"network\":\"Wifi-4A91D0\",\"password\":\"somePassword\"}" response on success will be the network name
```
Comment (elffjs) [inline README.md:29]: Modify this? Doesn't have to be in this PR.

File: main.go [hunk 9]
```diff
@@ -566,6 +632,72 @@ func main() {
...
+		setWifiResp, err := setWifiConnection(unitID, newWifiList)
+		if err != nil {
+			log.Printf("Failed to set wifi connection: %s", err)
+			return
+		}
+
+		if setWifiResp.Result {
+			log.Printf("Wifi Connection set successfully: %s", SSID)
+		} else {
+			log.Printf("Failed to set wifi connection: %s", err)
```
[the remark this answers is not in the document]
Comment (elffjs) [reply main.go:665]: I would still like to return an error here. I think we can merge this though.

**Review by elffjs — APPROVED** [review]

**Pull request merged**